### PR TITLE
Feat/add training scripts pipeline

### LINKED ISSUE #311: KG assets update pipeline
New papers and repositories are published and made available all the time. We need to include them in KG assets

### LINKED ISSUE #312: KG maintenance: a pipeline for retraining transformer and topic prediction models
The transformer model powering the KG recommender will require periodic retraining otherwise its quality will deteriorate over time. Similarly for the topic model. We need to define and develop the components necessary to monitor the need for retraining and to implement the retraining when needed.

### PR DESCRIPTION
**Pull request by marcvivancos**
Adds all the code necessary to update the data assets and to retrain the models

- Closes  #312 
- Closes #311 

The asset creation part can be automated, the model retraining part is freezed and manual atm.

### CODE REVIEW

File: kg/src/pipe/compute_topics.py [hunk 1]
```diff
@@ -1,10 +1,10 @@
 import pickle
 from pathlib import Path
-from typing import Union
 
 import numpy as np
 import pandas as pd
 import torch
+from pipe import utils
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/compute_topics.py:7]: I suggest we move our own module imports to the next section of imports separated from 3rd party library imports, in agreement with PEP-8. Otherwise we might have a new developer on a project wondering what library pipe is...

File: kg/src/pipe/README.md [hunk 2]
```diff
@@ -3,21 +3,48 @@
...
+```
+## Update and store data
+
+Once you have run the pipeline you can commit your current data using:
+```
+dvc commit
+```
+This will commit in dvc the current version of output data.
+
+After that you can store in an external storage:
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/README.md:39]: How do I get access to the external storage if I don't have access yet?

File: kg/src/pipe/compute_topics.py [hunk 3]
```diff
@@ -168,7 +126,7 @@ def load_vectorizer(filepath=filepath):
 
         # Load excluded topic indices
         self.excluded_topics_idx = np.array(
-            [int(x) for x in read_list_from_file(Path(filepath) / "excluded_topics_idx.txt")])
+            [int(x) for x in utils.read_list_from_file(Path(filepath) / "excluded_topics_idx.txt")])
 
     def vectorize_inputs(self):
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/compute_topics.py:131]: I think in load_for_inference mode, we don't need vectorize_inputs method. It uses self.train_texts and self.val_texts, which don't exist in the inference-only class, so I think we should remove this method. In inference mode we are simply loading the vectorizer from a saved file and not retraining it.

File: kg/src/pipe/download_datasets.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+import requests
+from google.cloud import storage
+
+DOWNLOAD_PATH = "inputs"
+
+PWC_PAPERS_URL = "https://paperswithcode.com/media/about/papers-with-abstracts.json.gz"
+PWC_LINKS_URL = "https://paperswithcode.com/media/about/links-between-papers-and-code.json.gz"
+
+ARXIV_JSON = "gs://arxiv-dataset/metadata-v5/arxiv-metadata-oai.json"
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/download_datasets.py:12]: What is gs://? Is this input or output filepath?

File: kg/src/pipe/download_datasets.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+CHUNK_SIZE = 8192
+
+
+def download_file(origin: str, dest_path: str, chunk_size: int):
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/download_datasets.py:17]: Add docstring, e.g.
"""
Downloads a compressed file with .gz extension from origin URL, decompresses it and stores to the specified destination directory.

Args:
    origin: (str) URL of the input file
    dest_path: (str) destination directory path at which to save the output (without the filename)
    chunk_size: (int) ??

Returns:
   (Path) filepath to which the file has been downloaded
"""

File: kg/src/pipe/download_datasets.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+            d = zlib.decompressobj(zlib.MAX_WBITS | 32)
+            for chunk in r.iter_content(chunk_size=chunk_size):
+                b = d.decompress(chunk)
+                f.write(b)
+
+    return local_path
+
+
+def download_arxiv_json(origin: str, dest_path: str):
+    print(f"Downloading arxiv json to: {dest_path}/arxiv-metadata-oai-snapshot.json")
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/download_datasets.py:36]: Replace hard-coded "arxiv-metadata-oai-snapshot.json" with:

fname = Path(origin).name
print(f"Downloading arxiv json to: {dest_path}/{fname}")

File: kg/src/pipe/download_datasets.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    client = storage.Client.create_anonymous_client()
+    dest = Path(dest_path, "arxiv-metadata-oai-snapshot.json")
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/download_datasets.py:38]: See comment above. I suggest replacing the hardcoded:

dest = Path(dest_path, "arxiv-metadata-oai-snapshot.json")

with:

dest = Path(dest_path, fname)

File: kg/src/pipe/train_transformer.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    print(f"Val data: {len(val_texts)} abstracts")
+    print(f"Test data: {len(test_texts)} abstracts")
+
+    # Tokenize texts with DistilBERT tokenizer
+    tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
+    tokenizer_args = dict(truncation=True, padding=True, return_tensors='pt',
+                          max_length=tokenizer.model_max_length)
+
+    # GPU benchmarking
+    train_encodings, time_tokenizer_train = tokenize(train_texts, tokenizer, tokenizer_args)
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/train_transformer.py:184]: This unpacking into a tuple with the second element being the time doesn't look correct if we are not decorating the function with `@return_with_time`. Does it work?

File: kg/src/pipe/train_transformer.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+        tuple:
+            batch_accuracy: (float)
+            (y_trues, y_preds): (tuple of list)
+    """
+    assert batch_labels.shape[0] == batch_outputs.logits.shape[
+        0], "Batch size must match between batch_outputs and batch_labels"
+    assert batch_labels.shape[1] == batch_outputs.logits.shape[
+        1], "Token number must match between batch_outputs and batch_labels"
+
+    MASK_TOKEN = -100
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/train_transformer.py:89]: We should replace this hard-coded `MASK_TOKEN` value with a call to one of the huggingface objects attributes/configs. Let me find where to access it from.

File: kg/src/pipe/train_transformer.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+    # Compute accuracy (train/val) before training
+    val_acc_before, (_, _) = run_accuracy_loop(model, val_loader, data_collator)
+    print(f"Val accuracy before training: {val_acc_before:.3f}")
+    train_acc_before, (_, _) = run_accuracy_loop(model, train_loader, data_collator)
+    print(f"Train accuracy before training: {train_acc_before:.3f}")
+
+    # Run training
+    train_loss, time_training = train_trainer(trainer=trainer)
+    val_loss, time_validation = evaluate_trainer(trainer=trainer, eval_dataset=val_dataset)
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/train_transformer.py:230]: Again, we are missing the `@return_with_wall_time` decorator in `train_trainer` and `evaluate_trainer` for this to work correctly.

File: kg/src/pipe/train_transformer.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+    print(f"Tokenization time (train): {time_tokenizer_train:3f}")
+    print(f"Tokenization time (val): {time_tokenizer_val:3f}")
+    print(f"Training time: {time_training:.3f}")
+
+    # Compute accuracy (train/val) after training
+    val_acc_after, (_, _) = run_accuracy_loop(model, val_loader, data_collator)
+    train_acc_after, (_, _) = run_accuracy_loop(model, train_loader, data_collator)
+
+    # Save and log the model
+    model.save_pretrained(MODEL_PATH)
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/train_transformer.py:241]: I would suggest checking that this path exists and is accessible before launching a long multi-GPU training

File: kg/src/pipe/train_transformer.py (hunk 9 above)
Comment (IGZVasjaUrbancic) [reply kg/src/pipe/train_transformer.py:89]: Ignore my previous comment. It is already hardcoded in the huggingface transformers library. 

While other DataCollators e.g. for Seq2Seq and for TokenClassification have a `.label_pad_token_id = -100` attribute, ours (`DataCollatorForLanguageModeling`) does not. (See https://github.com/huggingface/transformers/blob/master/src/transformers/data/data_collator.py)

We cannot do anything about it here.

File: kg/src/pipe/download_datasets.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+            # 32 is the header for gzip
+            d = zlib.decompressobj(zlib.MAX_WBITS | 32)
+            for chunk in r.iter_content(chunk_size=chunk_size):
+                b = d.decompress(chunk)
+                f.write(b)
+
+    return local_path
+
+
+def download_arxiv_json(origin: str, dest_path: str):
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/download_datasets.py:35]: Add docstring, e.g.
"""
Downloads a json file with .gz extension from origin URL, decompresses it and stores to the specified destination directory.

Args:
origin: (str) URL of the input file
dest_path: (str) destination directory path at which to save the output (without the filename)

Returns:
(None)
"""

File: kg/src/pipe/compute_vectors.py [hunk 13]
```diff
@@ -1,10 +1,10 @@
-import faiss
 from typing import Union
 
+import faiss
 import numpy as np
-import pandas as pd
 import torch
 import transformers
+from pipe import utils
 from pipe.data_checks.vectors import check_vector_size
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/compute_vectors.py:8]: I suggest moving our own library imports to a third block of imports separate from the 3rd-party library imports, in agreement with the PEP-8 convention

**Review by IGZVasjaUrbancic — COMMENTED** [review]
Great additions, thank you Marc!

I have two main comments:
- The `train_transformer.py` are missing the decorators to return running time, yet their usage implies that the time is returned. It doesn't seem like this will work as expected.
- Do we intend to include the code for retraining the topic model as well? Unless I missed it, we currently only have the inference variant of the model. For long-term maintenance, I think it would be good to add the topic model training code as well

File: kg/src/pipe/topics_model.py [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+        self.model = None
+        self.report_df = None
+        self.excluded_topics = None
+        self.excluded_topics_idx = None
+
+        if load_for_inference:
+            self.load_for_inference(filepath_saved_model)
+
+        else:
+            raise NotImplementedError
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/topics_model.py:143]: Is this still appropriate? It seems to me that this line will trip `train_topics_model` in `pipe/train_topics.py` (at line 38)

File: kg/src/pipe/topics_model.py [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+        item = {'X': self.X[idx],
+                'labels': self.labels[idx]}
+        return item
+
+    def __len__(self):
+        return len(self.X)
+
+
+class DNNModel(nn.Module):
+
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/topics_model.py:100]: Needs a docstring

File: kg/src/pipe/topics_model.py [hunk 16]
```diff
@@ -0,0 +1,328 @@
...
+        )
+
+        # Extract features using TFIDF vectorizer
+        vectorizer = TfidfVectorizer(strip_accents='unicode',
+                                     stop_words='english',
+                                     lowercase=True,
+                                     **VECTORIZER_ARGS)
+        X_train = vectorizer.fit_transform(
+            self.texts_train).toarray()  # Converting sparse matrix to array makes it easier to use downstream
+        X_val = vectorizer.transform(self.texts_val).toarray()
```
Comment (IGZVasjaUrbancic) [inline kg/src/pipe/topics_model.py:190]: `self.texts_train` and `self.texts_val` are not defined before usage
